### PR TITLE
Optimize AOTAutograd compiled forward wrapper

### LINKED ISSUE #122029: AOTAutograd has high fixed overheads
A few months ago [I looked at the fixed overheads of torch.compile](https://dev-discuss.pytorch.org/t/micro-optimizations-for-the-most-micro-of-benchmarks/1836) using [this microbenchmark](https://github.com/pytorch/pytorch/blob/6d9588a12b5834f29bd8970936749a9725e7f609/benchmarks/dynamo/microbenchmarks/overheads.py).

I wanted to give a quick update to that, in that since I realized I didn’t test things with `requires_grad=True` which triggers a bunch of extra wrapper code in AOTAutograd.

Here are, updated numbers from the last post with latest main.  Sadly, we have regressed 3.7us since last time.
- eager    4.6us
- compiled 14.0us

As an aside, if I enable `torch.inference_mode()` things get a bit better:
- eager    4.1us
- compiled 11.1us

Now on the other end, if I add `requires_grad=True` to [this line](https://github.com/pytorch/pytorch/blob/6d9588a12b5834f29bd8970936749a9725e7f609/benchmarks/dynamo/microbenchmarks/overheads.py#L14) of the microbenchmark, I get:
- eager    5.6us
- compiled 39.0us

Almost 2.8x worse than compile without the need to track gradients.  Note I am not calling backward, just triggering extra wrapper code to deal with backward if it does get called.  

So where is it coming from?  Here is the top of a cProfile (run 1000 times):
```
   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
     1000    0.010    0.000    0.026    0.000 jit_compile_runtime_wrappers.py:505(forward)
2000/1000    0.007    0.000    0.076    0.000 eval_frame.py:427(_fn)
    20000    0.007    0.000    0.007    0.000 function.py:346(__getattribute__)
     1000    0.005    0.000    0.033    0.000 {built-in method apply}
2000/1000    0.005    0.000    0.043    0.000 utils.py:105(call_func_at_runtime_with_args)
     1000    0.005    0.000    0.005    0.000 {built-in method torch.tensor}
     1000    0.005    0.000    0.055    0.000 runtime_wrappers.py:77(runtime_wrapper)
     1000    0.002    0.000    0.039    0.000 function.py:582(apply)
     2000    0.002    0.000    0.002    0.000 eval_frame.py:140(change)
     1000    0.001    0.000    0.004    0.000 <string>:2(guard)
     1000    0.001    0.000    0.003    0.000 clph2nks5ejst7jdpwldntojs4uhnqr57fddd5w6jpqaqposu7td.py:46(call)
```

It looks like the worst offender is coming form [jit_compile_runtime_wrappers.py](https://github.com/pytorch/pytorch/blob/6d9588a12b5834f29bd8970936749a9725e7f609/torch/_functorch/_aot_autograd/jit_compile_runtime_wrappers.py#L506).  As a quick experiment I wrote a patch that just skips that entire function:
```diff
diff --git a/torch/_functorch/_aot_autograd/jit_compile_runtime_wrappers.py b/torch/_functorch/_aot_autograd/jit_compile_runtime_wrappers.py
index ce087d44e25..e4328df32a9 100644
--- a/torch/_functorch/_aot_autograd/jit_compile_runtime_wrappers.py
+++ b/torch/_functorch/_aot_autograd/jit_compile_runtime_wrappers.py
@@ -504,6 +504,8 @@ def aot_dispatch_autograd(
 
         @staticmethod
         def forward(ctx, *deduped_flat_tensor_args):
+            return CompiledFunction.compiled_fw([*deduped_flat_tensor_args])[:1]
+
             args = deduped_flat_tensor_args
             if backward_state_indices:
                 bw_state = args[backward_state_indices[0]]
```

This improves things to:
- compiled 26.3us

Obviously we can’t land that, but it does show the maximum benefit that could be had by optimizing that function.  More than half of the overhead added by AOTAutograd is coming from that one function.  Looking over that function’s code, it seems like there are a lot of checks for various corner cases that will do nothing on most invocations.  In addition, nearly all of those corner case checks should be statically known at compile time.  One way to speed that up would be to codegen a custom wrapper function that only includes the corner cases that are actually relevant to the function being called.  Though some more basic trimming and code optimization would also likely help.

I don't plan to work on this, so this issue is up for grabs if someone is interested in some micro-optimizations.

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.16.0) (oldest at bottom):
* __->__ #186493

AOTAutograd already codegen'd the CompiledFunction.forward entry point, but the generated function still delegated every call to generic save/finalize helpers. For tiny compiled functions those helpers did mostly static or empty work on each invocation: slicing saved outputs, checking empty saved state, setting empty symint/opaque state, collecting non-differentiable outputs, and wrapping aliases based on compile-time metadata.

Move that static work into _codegen_compiled_forward. The generated forward now emits direct saved-tensor, symint, opaque-object, alias, unsafe-view, non-differentiable, and return handling for the specific compiled graph. The fresh compile and cache-hit paths now thread num_fw_outs_saved_for_bw so the codegen can recover the saved-output layout without calling the old generic helpers.

This follows the existing runtime-wrapper codegen direction instead of adding another runtime hook list or FX wrapper pass: it keeps the hot path straight-line Python and preserves existing assertions in generated form, including debug-only mutated-input and aliased-intermediate checks.

Fixes #122029

Generated by my agent

Benchmark Results

Command: python benchmarks/dynamo/microbenchmarks/overheads.py

Before: requires_grad=False compiled 14.1us; requires_grad=True compiled 26.5us; inference_mode compiled 13.7us.

After: requires_grad=False compiled 14.6us; requires_grad=True compiled 20.9us; inference_mode compiled 14.5us.

The target requires_grad=True compiled path improved by 5.6us, and the extra grad-enabled overhead over no-grad compiled dropped from 12.4us to 6.3us in these runs.

Test Plan

python -m py_compile torch/_functorch/_aot_autograd/runtime_wrappers.py torch/_functorch/_aot_autograd/graph_compile.py torch/_functorch/_aot_autograd/aot_autograd_result.py

Shimmed local test/functorch/test_aotdispatch.py -k compiled_forward: 11 tests OK

Shimmed local test/functorch/test_aotdispatch.py -k compiled_backward: 12 tests OK

Shimmed local test/inductor/test_cpu_repro.py -k test_compile_dynamic_grad_conv_transpose2d: 1 test OK

Shimmed local test/inductor/test_cpu_repro.py -k test_group_norm_backward_symint_divisible_channels: 1 test OK

Shimmed local test/inductor/test_cpu_repro.py -k test_backward_dynamic_item_from_size1_tensor: 1 test OK

python test/dynamo/test_aot_autograd_cache.py -k test_cache_hit: 12 tests OK

lintrunner -a

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @jataylo @azahed98

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I fixed the repeated AOTAutograd CI failures in test_codegen_output_alias and structured_trace: the generated forward wrapper now keeps the historical compiled_fn_wrapper artifact name while preserving the inlined fast path, and the structured trace expectation matches the resulting artifact order. I also added the missing topic: not user facing label. The DebertaV2ForMaskedLM entry in the previous Dr. CI run failed before Dynamo/AOTAutograd during eager Transformers import, so I left that unrelated benchmark-baseline issue for the fresh CI run to classify.

**Comment by jansel** [comment]
My agent says I rebased this onto current main and ported the forward specialization to PySourceBuilder while preserving graph-input view saves, intermediate-view detaches, external-object snapshots, current custom-class checks, and copied backward contexts. Targeted CPU, CUDA, cache, structured-trace, and compiled-autograd tests plus lintrunner pass; current-main benchmarking improves the requires_grad=True compiled path from 21.0us to 15.9us.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says this PR is being closed as part of a cleanup of open agentic PRs to focus review bandwidth on a smaller set; the linked issue remains open.
